Mathlib/LinearAlgebra/Dimension/RankDecomposition.lean

```diff
@@ -0,0 +1,214 @@
+/-
+Copyright (c) 2025 __________________. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: __________________
+-/
+import Mathlib.LinearAlgebra.Dimension.LinearMap
+import Mathlib.LinearAlgebra.FiniteDimensional.Lemmas
+import Mathlib.Order.CompletePartialOrder
+
+/-!
+# Rank decomposition and bases adapted to a linear map
+
+This file develops a small API around a linear map `f : V →ₗ[K] W` that chooses a complement
+of `ker f`, builds bases compatible with the direct-sum decomposition
+`V = ker f ⊕ kerComplement f`, and produces a rank-normal form for the matrix of `f`:
+after choosing suitable bases on domain and codomain, the matrix of `f` is a block matrix
+`fromBlocks 1 0 0 0` (an identity of size `rank f` in the top-left block and zeros elsewhere).
+
+## Main definitions
+
+* `LinearMap.kerComplement` :
+  a fixed choice of a subspace complementary to `ker f`.
+
+* `LinearMap.prodEquivOfkerComplement` :
+  a linear equivalence `(kerComplement f × ker f) ≃ₗ[K] V`
+  induced by the complementary decomposition.
+
+* `LinearMap.decomposition_basis` :
+  a basis of `V` obtained by transporting the product of the canonical bases on
+  `kerComplement f` and `ker f` along `prodEquivOfkerComplement`.
+
+* `LinearMap.ker_complement_restriction` :
+  the restriction of `f` to `kerComplement f` (include, then apply `f`).
+
+* `LinearMap.ker_complement_basis_image` :
+  the image in `W` of the canonical basis of `kerComplement f` under the restriction map.
+
+* `LinearMap.range_decomposition_basis` :
+  a basis of `W` built by extending the previous image (which spans `range f`)
+  via `Basis.sumExtend`.
```

Comment:
sounds like you followed a textbook's low level proof of the rank nullity equality, which might not be the best pathway for lean formalisation